tests/python/relay/test_vm_serialization.py

```diff
@@ -0,0 +1,313 @@
+    tup = relay.var('tup', type_annotation=ttype)
+    f = relay.Function([tup], relay.TupleGetItem(tup, 1))
+    i_data = np.random.rand(41).astype('float32')
+    j_data = np.random.rand(10).astype('float32')
+
+    vm = create_vm(f)
+    ser = serializer.Serializer(vm)
+    code, lib = ser.serialize()
+    deser = deserializer.Deserializer(code, lib)
+    des_vm = deser.deserialize()
+
+    result = veval(des_vm, (i_data, j_data))
+    tvm.testing.assert_allclose(result.asnumpy(), j_data)
+
+
+def test_adt_list():
+    mod = relay.Module()
+    p = Prelude(mod)
+
+    l1 = p.cons(relay.const(1), p.nil())
+    l21 = p.cons(relay.const(2), l1)
+    l321 = p.cons(relay.const(3), l21)
+
+    f = relay.Function([], l321)
+    mod["main"] = f
+
+    vm = create_vm(mod)
+    ser = serializer.Serializer(vm)
+    code, lib = ser.serialize()
+    deser = deserializer.Deserializer(code, lib)
+    des_vm = deser.deserialize()
+
+    result = veval(des_vm)
+    assert len(result) == 2
+    assert len(result[1]) == 2
+    assert len(result[1][1]) == 2
+    res = []
+    res.append(result[0].asnumpy().tolist())
+    res.append(result[1][0].asnumpy().tolist())
+    res.append(result[1][1][0].asnumpy().tolist())
+    tvm.testing.assert_allclose(res, np.array([3, 2, 1]))
+
+
+def test_adt_compose():
+    mod = relay.Module()
+    p = Prelude(mod)
+
+    compose = p.compose
+
+    # add_one = fun x -> x + 1
+    sb = relay.ScopeBuilder()
+    x = relay.var('x', 'float32')
+    x1 = sb.let('x1', x)
+    xplusone = x1 + relay.const(1.0, 'float32')
+    sb.ret(xplusone)
+    body = sb.get()
+    add_one = relay.GlobalVar("add_one")
+    add_one_func = relay.Function([x], body)
+
+    # add_two = compose(add_one, add_one)
+    sb = relay.ScopeBuilder()
+    y = relay.var('y', 'float32')
+    add_two_func = sb.let('add_two', compose(add_one_func, add_one_func))
+    add_two_res = add_two_func(y)
+    sb.ret(add_two_res)
+    add_two_body = sb.get()
+
+    mod[add_one] = add_one_func
+
+    f = relay.Function([y], add_two_body)
+    mod["main"] = f
+
+    vm = create_vm(mod)
+    ser = serializer.Serializer(vm)
+    code, lib = ser.serialize()
+    deser = deserializer.Deserializer(code, lib)
+    des_vm = deser.deserialize()
+
+    x_data = np.array(np.random.rand()).astype('float32')
+    result = veval(des_vm, x_data)
+
+    tvm.testing.assert_allclose(result.asnumpy(), x_data + 2.0)
+
+
+def test_closure():
+    x = relay.var('x', shape=())
+    y = relay.var('y', shape=())
+    f = relay.Function([x], x + y)
+    ff = relay.Function([y], f)
+    clo = ff(relay.const(1.0))
+    main = clo(relay.const(2.0))
+
+    vm = create_vm(main)
+    ser = serializer.Serializer(vm)
+    code, lib = ser.serialize()
+    deser = deserializer.Deserializer(code, lib)
+    des_vm = deser.deserialize()
+
+    res = veval(des_vm)
+    tvm.testing.assert_allclose(res.asnumpy(), 3.0)
+
+
+if __name__ == "__main__":
```

Comment:
Can you write a loop over all the testing models load them into vm and serialize them and ensure they produce the same result after deserialization?

Reply:
May be just 1 or 2? It may make the testing too slow have all them in the unit test.